### PR TITLE
build(tree): Make compatible with moduleResolution: node16

### PR DESCRIPTION
**Pull request by tylerbutler**
Updates the tree package to build ESM using tsc-multi, create valid ESM types, and adds an exports field and testing using arethetypeswrong.

This change also required more changes to tsc-multi to handle the types tree uses. Those changes are included here by updating the patch.

### CODE REVIEW

File: packages/dds/tree/src/domains/schemaBuilder.ts [hunk 1]
```diff
@@ -273,27 +273,27 @@ export class SchemaBuilder<
 	/**
 	 * {@link leaf.number}
 	 */
-	public readonly number = leaf.number;
+	public readonly number: typeof leaf.number = leaf.number;
```
Comment (alexvy86) [inline packages/dds/tree/src/domains/schemaBuilder.ts:276]: Some explicit types added where they were simple, to reduce the number of dynamic imports in `d.?ts` files.

File: packages/dds/tree/src/feature-libraries/default-schema/defaultFieldKinds.ts [hunk 2]
```diff
@@ -59,16 +60,17 @@ const optionalIdentifier = "Optional";
...
-	optionalIdentifier,
-	Multiplicity.Optional,
-	optionalChangeHandler,
-	(types, other) =>
-		(other.kind.identifier === sequence.identifier ||
-			other.kind.identifier === optionalIdentifier) &&
-		allowsTreeSchemaIdentifierSuperset(types, other.types),
-	new Set([]),
-);
+export const optional: FieldKindWithEditor<OptionalFieldEditor, Multiplicity.Optional, "Optional"> =
```
Comment (alexvy86) [inline packages/dds/tree/src/feature-libraries/default-schema/defaultFieldKinds.ts:63]: Some explicit types added where they were simple, to reduce the number of dynamic imports in `d.?ts` files (same below in this file).

File: packages/dds/tree/src/feature-libraries/typed-schema/typedTreeSchema.ts [hunk 3]
```diff
@@ -328,7 +328,10 @@ export class TreeFieldSchema<
 	/**
 	 * Schema for a field which must always be empty.
 	 */
-	public static readonly empty = TreeFieldSchema.create(FieldKinds.forbidden, []);
+	public static readonly empty: TreeFieldSchema<Forbidden, readonly []> = TreeFieldSchema.create(
```
Comment (alexvy86) [inline packages/dds/tree/src/feature-libraries/typed-schema/typedTreeSchema.ts:331]: Some explicit types added where they were simple, to reduce the number of dynamic imports in `d.?ts` files.

File: packages/dds/tree/src/feature-libraries/optional-field/index.ts [hunk 4]
```diff
@@ -8,6 +8,7 @@ export { OptionalChangeset, RegisterId } from "./optionalFieldChangeTypes";
 export {
 	optionalChangeHandler,
 	optionalFieldEditor,
+	OptionalFieldEditor,
```
Comment (alexvy86) [inline packages/dds/tree/src/feature-libraries/optional-field/index.ts:11]: This was done so I could add explicit types to some things, following what sequence field already did in terms of exports.

File: packages/dds/tree/src/core/schema-stored/storedSchemaRepository.ts [hunk 5]
```diff
@@ -54,13 +54,17 @@ export interface MutableTreeStoredSchema extends TreeStoredSchemaSubscription {
 	apply(newSchema: TreeStoredSchema): void;
 }
 
+type eventsEmitterType = ISubscribable<SchemaEvents> &
```
Comment (alexvy86) [inline packages/dds/tree/src/core/schema-stored/storedSchemaRepository.ts:57]: Some explicit types added where they were simple, to reduce the number of dynamic imports in `d.?ts` files.

File: experimental/dds/tree/package.json [hunk 6]
```diff
@@ -28,17 +52,17 @@
...
 		"test": "npm run test:mocha",
-		"test:benchmark:report": "mocha \"dist/**/*.tests.js\" --node-option unhandled-rejections=strict,expose-gc --exit --perfMode --fgrep @Benchmark -r node_modules/@fluidframework/mocha-test-setup --reporter @fluid-tools/benchmark/dist/MochaReporter.js --timeout 60000",
+		"test:benchmark:report": "mocha \"dist/**/*.tests.*js\" --node-option unhandled-rejections=strict,expose-gc --exit --perfMode --fgrep @Benchmark -r node_modules/@fluidframework/mocha-test-setup --reporter @fluid-tools/benchmark/dist/MochaReporter.js --timeout 60000",
 		"test:coverage": "c8 npm test",
-		"test:mocha": "mocha \"dist/**/*.tests.js\" --exit -r node_modules/@fluidframework/mocha-test-setup",
+		"test:mocha": "mocha \"dist/**/*.tests.*js\" --exit -r node_modules/@fluidframework/mocha-test-setup",
 		"test:mocha:verbose": "cross-env FLUID_TEST_VERBOSE=1 npm run test:mocha",
 		"test:stress": "cross-env FUZZ_TEST_COUNT=10 FUZZ_STRESS_RUN=true mocha \"dist/**/*.fuzz.tests.js\" --exit -r @fluidframework/mocha-test-setup",
-		"tsc": "tsc"
+		"tsc": "tsc-multi --config ../../../common/build/build-common/tsc-multi.cjs.json"
```
Comment (tylerbutler) [inline experimental/dds/tree/package.json:65]: This was the old style, so I'll update it to use tsc directly for CJS like we do elsewhere.

File: experimental/dds/tree/package.json [hunk 7]
```diff
@@ -90,7 +115,9 @@
 		"mocha-multi-reporters": "^1.5.1",
 		"moment": "^2.21.0",
 		"prettier": "~3.0.3",
+		"renamer": "^4.0.0",
```
Comment (tylerbutler) [inline experimental/dds/tree/package.json:118]: remove

File: experimental/dds/tree/package.json [hunk 8]
```diff
@@ -11,14 +11,38 @@
...
+	"main": "dist/index.cjs",
+	"module": "lib/index.mjs",
 	"types": "dist/index.d.ts",
 	"scripts": {
 		"build": "fluid-build . --task build",
 		"build:compile": "fluid-build . --task compile",
 		"build:docs": "api-extractor run --local",
-		"build:esnext": "tsc --project ./tsconfig.esnext.json",
+		"build:esnext": "tsc-multi --config ../../../common/build/build-common/tsc-multi.esm.json",
+		"build:rename-types": "renamer \"lib/**\" -f .d.ts -r .d.mts --force",
```
Comment (tylerbutler) [inline experimental/dds/tree/package.json:44]: remove

File: experimental/framework/tree-react-api/tsc-multi.test.json
Comment (tylerbutler) [inline experimental/framework/tree-react-api/tsc-multi.test.json:1]: remove this file

File: experimental/dds/tree/package.json [hunk 9]
```diff
@@ -11,14 +11,37 @@
...
+			},
+			"require": {
+				"types": "./dist/EditLog.d.ts",
+				"default": "./dist/EditLog.js"
+			}
+		}
+	},
 	"main": "dist/index.js",
-	"module": "lib/index.js",
+	"module": "lib/index.mjs",
```
Comment (jason-ha) [inline experimental/dds/tree/package.json:37]: Should just be removed - module is not supported and never will be

File: packages/dds/tree/package.json [hunk 10]
```diff
@@ -12,17 +12,19 @@
 	"author": "Microsoft and contributors",
 	"sideEffects": false,
 	"main": "dist/index.js",
-	"module": "lib/index.js",
+	"module": "lib/index.mjs",
```
Comment (jason-ha) [inline packages/dds/tree/package.json:15]: Should just be removed - module is not supported

File: packages/dds/tree/package.json [hunk 11]
```diff
@@ -12,17 +12,19 @@
...
+		"api-extractor:esnext": "api-extractor run --config ./api-extractor-esm.json",
 		"bench": "mocha --timeout 999999 --perfMode --parentProcess --fgrep @Benchmark --reporter @fluid-tools/benchmark/dist/MochaReporter.js",
 		"bench:profile": "mocha --v8-prof --v8-logfile=profile.log --v8-no-logfile-per-isolate --timeout 999999 --perfMode --fgrep @Benchmark --reporter @fluid-tools/benchmark/dist/MochaReporter.js && node --prof-process profile.log > profile.txt && rimraf profile.log && echo See results in profile.txt",
 		"build": "fluid-build . --task build",
 		"build:commonjs": "fluid-build . --task commonjs",
 		"build:compile": "fluid-build . --task compile",
 		"build:docs": "api-extractor run --local",
-		"build:esnext": "tsc --project ./tsconfig.esnext.json",
-		"build:genver": "gen-version",
+		"build:esnext": "tsc-multi --config ../../../common/build/build-common/tsc-multi.esm.json",
```
Comment (jason-ha) [inline packages/dds/tree/package.json:27]: I think we should prefer ESM first general but at least for `tree`. Make sure we see all of the issues with ESM without filtering.

File: packages/dds/tree/package.json [hunk 12]
```diff
@@ -12,17 +12,19 @@
 	"author": "Microsoft and contributors",
 	"sideEffects": false,
 	"main": "dist/index.js",
```
Comment (jason-ha) [inline packages/dds/tree/package.json:14]: "type" should be set (prefer "module" with esm as first class.)
and "exports" should be added.

Comment (tylerbutler) [reply packages/dds/tree/package.json:14]: Agree on removing the main/module field ultimately, but I'd like to do it separately, and ideally for all packages in the release group. Also, are you sure that removing these is not a breaking change? I thought at least webpack used the module field, but haven't yet confirmed whether removing it breaks our own examples. That's the other (minor) reason I want to do it all at once - so I can make only one "breaking change" note.

File: packages/dds/tree/api-report/tree.api.md [hunk 13]
```diff
@@ -4,6 +4,8 @@
 
 ```ts
 
+import { FieldKind as FieldKind_2 } from './schemaTypes.js';
+import { FieldSchema as FieldSchema_2 } from './schemaTypes.js';
 import { FluidObject } from '@fluidframework/core-interfaces';
```
Comment (jason-ha) [inline packages/dds/tree/api-report/tree.api.md:9]: These are not valid which means the API extractor generated .d.ts files are probably incorrect.

Comment (jason-ha) [reply packages/dds/tree/api-report/tree.api.md:9]: Here is an example log when I had such gunk: [https://dev.azure.com/fluidframework/public/_build/results?buildId=228628&view=logs&j=3dc8fd7e-4368-5a92-293e-d53cefc8c4b3&t=bfe27d5f-9219-57c0-0d67-1b2946a32921](https://dev.azure.com/fluidframework/public/_build/results?buildId=228628&view=logs&j=3dc8fd7e-4368-5a92-293e-d53cefc8c4b3&t=bfe27d5f-9219-57c0-0d67-1b2946a32921)

File: packages/dds/tree/src/domains/schemaBuilder.ts (hunk 1 above)
Comment (jason-ha) [reply packages/dds/tree/src/domains/schemaBuilder.ts:280]: We shouldn't have to make any source changes to get rollups and generate API reports.
Alternate solution for tree #19141

File: packages/dds/tree/api-report/tree.api.md (hunk 13 above)
Comment (tylerbutler) [reply packages/dds/tree/api-report/tree.api.md:9]: It's worrisome to me that in the example you linked, the tools involved were only tsc and api-extractor. Does that imply that api-extractor itself doesn't like the Tree types? Maybe we should try updating api-extractor? Looks like the recent minor release added support for TypeScript 5.3 - https://github.com/microsoft/rushstack/blob/main/apps/api-extractor/CHANGELOG.md#7390

@Josmithr FYI.

Comment (jason-ha) [reply packages/dds/tree/api-report/tree.api.md:9]: I was guessing it isn't a Tree type thing but something with moduleResolution support.

I did a test with updated api-extractor for my branch and results were not great.
First I tried exactly as I had things and then with tree typescript set to 5.3.3 which the changelog said is supported:
```log
@fluidframework/tree: error during command 'api-extractor run --local --config ./api-extractor-lint.json' (exit code 1)
@fluidframework/tree: api-extractor 7.39.1  - https://api-extractor.com/
@fluidframework/tree: 
@fluidframework/tree: 
@fluidframework/tree: 
@fluidframework/tree: ERROR: Error parsing /home/jasonha/ff/FluidFramework/packages/dds/tree/api-extractor-lint.json:
@fluidframework/tree: The "bundledPackages" list contains an invalid package name: "@fluidframework/.*"
[ 8/12] x @fluidframework/tree: api-extractor run --local --config ./api-extractor-lint.json - 0.327s
@fluidframework/tree: error during command 'api-extractor run --config ./api-extractor-cjs.json' (exit code 1)
@fluidframework/tree: api-extractor 7.39.1  - https://api-extractor.com/
@fluidframework/tree: 
@fluidframework/tree: Analysis will use the bundled TypeScript version 5.1.6
@fluidframework/tree: *** The target project appears to use TypeScript 5.3.3 which is newer than the bundled compiler engine; consider upgrading API Extractor.
@fluidframework/tree: 
@fluidframework/tree: 
@fluidframework/tree: ERROR: program.getResolvedModule is not a function
[ 9/12] x @fluidframework/tree: api-extractor run --config ./api-extractor-cjs.json - 2.742s
@fluidframework/tree: error during command 'api-extractor run --local --config ./api-extractor-esm.json' (exit code 1)
@fluidframework/tree: api-extractor 7.39.1  - https://api-extractor.com/
@fluidframework/tree: 
@fluidframework/tree: Analysis will use the bundled TypeScript version 5.1.6
@fluidframework/tree: *** The target project appears to use TypeScript 5.3.3 which is newer than the bundled compiler engine; consider upgrading API Extractor.
@fluidframework/tree: 
@fluidframework/tree: 
@fluidframework/tree: ERROR: program.getResolvedModule is not a function
[10/12] x @fluidframework/tree: api-extractor run --local --config ./api-extractor-esm.json - 2.726s
```

Comment (tylerbutler) [reply packages/dds/tree/api-report/tree.api.md:9]: These errors might be related to the manual api-extractor patch that @Josmithr has been managing. Presumably that patch will need to be regenerated.

Comment (jason-ha) [reply packages/dds/tree/api-report/tree.api.md:9]: Good call out. The patch is listed as for a specific version but might not be. I then also found that we were forcing 5.1.6 into api-extractor per pnpm override. When I removed those two things the main runs worked. The lint case still complains about malformed bundledPackages.
No change on the results for the different module resolutions and bad results.
I feel like I've seen such things in other packages, but maybe it was always when playing with module resolution. I think there was some checked in at some point and then something corrected it later.

Comment (jason-ha) [reply packages/dds/tree/api-report/tree.api.md:9]: Ha, ha. The patch is specifically about making `bundledPackages` support regex.

**Pull request closed without merging**